File: python/packages/core/src/matimo/core/skill_content_parser.py
"""
Skill Content Parser — Markdown heading-based section chunking.

Mirrors: packages/core/src/core/skill-content-parser.ts

Breaks skill bodies into structured sections so agents load only the parts
they need instead of dumping the entire SKILL.md into context.
"""
from __future__ import annotations

import math
import re
from dataclasses import dataclass, field

from matimo.core.models import SkillContentOptions, SkillSection


@dataclass
class ParsedSkillContent:
    """Result of parsing a skill body into sections."""

    preamble: str = ""
    preamble_tokens: int = 0
    sections: list[SkillSection] = field(default_factory=list)
    total_tokens: int = 0
    index: dict[str, SkillSection] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

_HEADING_RE = re.compile(r"^(#{1,6})\s+(.+)$")


def _parse_heading(line: str) -> tuple[int, str] | None:
    """Parse a Markdown ATX heading. Returns (level, text) or None."""
    if not line.startswith("#"):
        return None
    m = _HEADING_RE.match(line)
    if m is None:
        return None
    return len(m.group(1)), m.group(2).strip()


def _estimate_tokens(text: str) -> int:
    """Rough heuristic: 1 token ≈ 0.75 words for English text."""
    if not text:
        return 0
    word_count = len(text.split())
    return math.ceil(word_count / 0.75)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def parse_skill_sections(body: str) -> ParsedSkillContent:
    """
    Parse a Markdown skill body into a tree of sections.

    Lightweight heading-based parser that splits on ATX headings (``# …``).
    Handles fenced code blocks so that ``#`` inside code is not treated as a
    heading.
    """
    if not body or not body.strip():
        return ParsedSkillContent()

    lines = body.split("\n")
    index: dict[str, SkillSection] = {}

    # Gather raw segments
    segments: list[dict] = []
    current: dict | None = None
    preamble_lines: list[str] = []
    in_code_block = False

    for line in lines:
        stripped = line.lstrip()
        if stripped.startswith("```"):
            in_code_block = not in_code_block

        if not in_code_block:
            heading = _parse_heading(line)
            if heading is not None:
                if current is not None:
                    segments.append(current)
                current = {
                    "heading": heading[1],
                    "level": heading[0],
                    "content_lines": [],
                }
                continue

        if current is not None:
            current["content_lines"].append(line)
        else:
            preamble_lines.append(line)

    if current is not None:
        segments.append(current)

    preamble = "\n".join(preamble_lines).strip()
    preamble_tokens = _estimate_tokens(preamble)

    # Build tree from flat list using a stack
    top_sections: list[SkillSection] = []
    stack: list[SkillSection] = []

    for seg in segments:
        content = "\n".join(seg["content_lines"]).strip()
        section = SkillSection(
            heading=seg["heading"],
            level=seg["level"],
            content=content,
            token_estimate=_estimate_tokens(content) + _estimate_tokens(seg["heading"]),
            children=[],
            path=seg["heading"],
        )

        # Pop stack until we find a parent with a lower level
        while stack and stack[-1].level >= seg["level"]:
            stack.pop()

        if stack:
            parent = stack[-1]
            section.path = f"{parent.path}.{seg['heading']}"
            parent.children.append(section)
        else:
            top_sections.append(section)

        index[section.path.lower()] = section
        stack.append(section)

    # Calculate total tokens (recursive)
    def _total(s: SkillSection) -> int:
        return s.token_estimate + sum(_total(c) for c in s.children)

    total_tokens = preamble_tokens + sum(_total(s) for s in top_sections)

    return ParsedSkillContent(
        preamble=preamble,
        preamble_tokens=preamble_tokens,
        sections=top_sections,
        total_tokens=total_tokens,
        index=index,
    )


def extract_skill_content(
    parsed: ParsedSkillContent,
    options: SkillContentOptions | None = None,
) -> str:
    """
    Selectively extract content from a parsed skill body.

    Key function for context management — agents call this to get only the
    sections they need instead of the entire SKILL.md.
    """
    if options is None:
        options = SkillContentOptions()

    parts: list[str] = []
    current_tokens = 0

    def within_budget(additional: int) -> bool:
        if options.max_tokens is None:
            return True
        return current_tokens + additional <= options.max_tokens

    def render_section(section: SkillSection, depth: int) -> str:
        hashes = "#" * section.level
        result = f"{hashes} {section.heading}\n\n{section.content}"
        if options.max_depth is None or depth < options.max_depth:
            for child in section.children:
                result += "\n\n" + render_section(child, depth + 1)
        return result

    # Add preamble
    if options.include_preamble and parsed.preamble:
        tokens = parsed.preamble_tokens
        if within_budget(tokens):
            parts.append(parsed.preamble)
            current_tokens += tokens

    # If specific sections requested, find and include only those
    if options.sections:
        for requested in options.sections:
            lower = requested.lower()
            found = parsed.index.get(lower)
            if not found:
                # Partial match
                for key, section in parsed.index.items():
                    if lower in key:
                        found = section
                        break

            if found:
                rendered = render_section(found, 1)
                tokens = _estimate_tokens(rendered)
                if within_budget(tokens):
                    parts.append(rendered)
                    current_tokens += tokens
    else:
        # Include all sections
        for section in parsed.sections:
            rendered = render_section(section, 1)
            tokens = _estimate_tokens(rendered)
            if within_budget(tokens):
                parts.append(rendered)
                current_tokens += tokens
            else:
                break

    return "\n\n".join(parts)


def list_skill_sections(
    parsed: ParsedSkillContent,
) -> list[dict[str, object]]:
    """
    Get a flat list of all section headings with their token costs.
    Useful for agents to decide which sections to load.
    """
    result: list[dict[str, object]] = []

    def walk(section: SkillSection) -> None:
        result.append(
            {
                "path": section.path,
                "level": section.level,
                "token_estimate": section.token_estimate,
            }
        )
        for child in section.children:
            walk(child)

    for section in parsed.sections:
        walk(section)

    return result

File: typescript/packages/core/src/core/skill-content-parser.ts
/**
 * Skill Content Parser — Markdown AST-based section chunking
 *
 * Breaks skill bodies into structured sections so agents load only the parts
 * they need instead of dumping the entire SKILL.md into context.
 *
 * Uses lightweight heading-based parsing (no external Markdown AST library
 * needed) to produce a tree of sections with token-count estimates.
 */

/**
 * A single section of a skill body, parsed from Markdown headings.
 */
export interface SkillSection {
  /** Heading text (e.g., "Error Handling") */
  heading: string;
  /** Heading level (1-6) */
  level: number;
  /** Raw Markdown content under this heading (excluding sub-headings) */
  content: string;
  /** Approximate token count (words ÷ 0.75 — conservative estimate) */
  tokenEstimate: number;
  /** Nested sub-sections */
  children: SkillSection[];
  /** Dot-path for addressing (e.g., "Error Handling.Rate Limits") */
  path: string;
}

/**
 * Result of parsing a skill body into sections.
 */
export interface ParsedSkillContent {
  /** Top-level intro content before any heading */
  preamble: string;
  /** Token estimate for preamble */
  preambleTokens: number;
  /** All top-level sections (with nested children) */
  sections: SkillSection[];
  /** Total token estimate for the entire body */
  totalTokens: number;
  /** Flat index: heading path → section reference (for fast lookup) */
  index: Map<string, SkillSection>;
}

/**
 * Options for retrieving skill content with selective loading.
 */
export interface SkillContentOptions {
  /** Only return sections matching these heading paths (case-insensitive partial match) */
  sections?: string[];
  /** Maximum total tokens to return (truncates from the end) */
  maxTokens?: number;
  /** Include the preamble (default: true) */
  includePreamble?: boolean;
  /** Depth limit for section inclusion (1 = top-level only, 2 = include children, etc.) */
  maxDepth?: number;
}

// ─── Helper: Check if line is a Markdown heading using safe string operations ─────────
// Avoids regex to prevent ReDoS on malicious input with many spaces
function parseHeading(line: string): { level: number; heading: string } | null {
  if (!line.startsWith('#')) return null;

  let level = 0;
  for (let i = 0; i < Math.min(6, line.length); i++) {
    if (line[i] === '#') level++;
    else break;
  }

  if (level === 0 || level > 6) return null;

  // Check that after the hashes, there's whitespace
  if (level >= line.length || !/\s/.test(line[level])) return null;

  // Extract heading text after the hashes and skip leading whitespace
  const heading = line.substring(level).trimStart();
  return { level, heading };
}

/**
 * Estimate token count from text.
 * Rough heuristic: 1 token ≈ 0.75 words for English text.
 * Conservative to avoid under-counting.
 */
function estimateTokens(text: string): number {
  if (!text) return 0;
  const wordCount = text.split(/\s+/).filter(Boolean).length;
  return Math.ceil(wordCount / 0.75);
}

/**
 * Parse a Markdown skill body into a tree of sections.
 *
 * This is a lightweight parser that splits on ATX headings (lines starting
 * with #). It does NOT handle:
 * - Setext headings (underline style)
 * - Headings inside code blocks (these are treated as content)
 *
 * For SKILL.md files (which follow a consistent format), this is sufficient.
 */
export function parseSkillSections(body: string): ParsedSkillContent {
  if (!body || body.trim().length === 0) {
    return {
      preamble: '',
      preambleTokens: 0,
      sections: [],
      totalTokens: 0,
      index: new Map(),
    };
  }

  const lines = body.split('\n');
  const index = new Map<string, SkillSection>();

  // Collect raw segments: [{heading, level, contentLines}]
  interface RawSegment {
    heading: string;
    level: number;
    contentLines: string[];
  }

  const segments: RawSegment[] = [];
  let currentSegment: RawSegment | null = null;
  const preambleLines: string[] = [];
  let inCodeBlock = false;

  for (const line of lines) {
    // Track fenced code blocks to avoid treating # inside them as headings
    if (line.trimStart().startsWith('```')) {
      inCodeBlock = !inCodeBlock;
    }

    if (!inCodeBlock) {
      const headingMatch = parseHeading(line);
      if (headingMatch) {
        // Flush previous segment
        if (currentSegment) {
          segments.push(currentSegment);
        }
        currentSegment = {
          heading: headingMatch.heading,
          level: headingMatch.level,
          contentLines: [],
        };
        continue;
      }
    }

    if (currentSegment) {
      currentSegment.contentLines.push(line);
    } else {
      preambleLines.push(line);
    }
  }

  // Flush last segment
  if (currentSegment) {
    segments.push(currentSegment);
  }

  const preamble = preambleLines.join('\n').trim();
  const preambleTokens = estimateTokens(preamble);

  // Build tree from flat list of segments using a stack
  const topSections: SkillSection[] = [];
  const stack: SkillSection[] = [];

  for (const seg of segments) {
    const content = seg.contentLines.join('\n').trim();
    const section: SkillSection = {
      heading: seg.heading,
      level: seg.level,
      content,
      tokenEstimate: estimateTokens(content) + estimateTokens(seg.heading),
      children: [],
      path: seg.heading,
    };

    // Pop stack until we find a parent with a lower level
    while (stack.length > 0 && stack[stack.length - 1].level >= seg.level) {
      stack.pop();
    }

    if (stack.length > 0) {
      const parent = stack[stack.length - 1];
      section.path = `${parent.path}.${seg.heading}`;
      parent.children.push(section);
    } else {
      topSections.push(section);
    }

    index.set(section.path.toLowerCase(), section);
    stack.push(section);
  }

  // Calculate total tokens (recursive)
  function totalTokensOf(section: SkillSection): number {
    return section.tokenEstimate + section.children.reduce((sum, c) => sum + totalTokensOf(c), 0);
  }

  const totalTokens = preambleTokens + topSections.reduce((sum, s) => sum + totalTokensOf(s), 0);

  return {
    preamble,
    preambleTokens,
    sections: topSections,
    totalTokens,
    index,
  };
}

/**
 * Selectively extract content from a parsed skill body.
 *
 * This is the key function for context management — instead of dumping the
 * entire SKILL.md into the LLM's context window, agents call this to get
 * only the sections they need.
 *
 * @example
 * // Get only the error handling section
 * extractSkillContent(parsed, { sections: ['Error Handling'] })
 *
 * @example
 * // Get top-level overview only (no sub-sections), max 500 tokens
 * extractSkillContent(parsed, { maxDepth: 1, maxTokens: 500 })
 */
export function extractSkillContent(
  parsed: ParsedSkillContent,
  options: SkillContentOptions = {}
): string {
  const { sections: requestedSections, maxTokens, includePreamble = true, maxDepth } = options;

  const parts: string[] = [];
  let currentTokens = 0;

  // Helper: check if we've exceeded the token budget
  function withinBudget(additional: number): boolean {
    if (maxTokens === undefined) return true;
    return currentTokens + additional <= maxTokens;
  }

  // Helper: render a section to Markdown
  function renderSection(section: SkillSection, depth: number): string {
    const hashes = '#'.repeat(section.level);
    let result = `${hashes} ${section.heading}\n\n${section.content}`;

    if (maxDepth === undefined || depth < maxDepth) {
      for (const child of section.children) {
        result += '\n\n' + renderSection(child, depth + 1);
      }
    }

    return result;
  }

  // Add preamble
  if (includePreamble && parsed.preamble) {
    const tokens = parsed.preambleTokens;
    if (withinBudget(tokens)) {
      parts.push(parsed.preamble);
      currentTokens += tokens;
    }
  }

  // If specific sections requested, find and include only those
  if (requestedSections && requestedSections.length > 0) {
    for (const requested of requestedSections) {
      const lower = requested.toLowerCase();

      // Try exact match first, then partial match
      let found = parsed.index.get(lower);
      if (!found) {
        // Partial match: find first section whose path contains the query
        for (const [key, section] of parsed.index) {
          if (key.includes(lower)) {
            found = section;
            break;
          }
        }
      }

      if (found) {
        const rendered = renderSection(found, 1);
        const tokens = estimateTokens(rendered);
        if (withinBudget(tokens)) {
          parts.push(rendered);
          currentTokens += tokens;
        }
      }
    }
  } else {
    // Include all sections (respecting maxDepth and maxTokens)
    for (const section of parsed.sections) {
      const rendered = renderSection(section, 1);
      const tokens = estimateTokens(rendered);
      if (withinBudget(tokens)) {
        parts.push(rendered);
        currentTokens += tokens;
      } else {
        break; // Stop adding sections once we exceed budget
      }
    }
  }

  return parts.join('\n\n');
}

/**
 * Get a flat list of all section headings with their token costs.
 * Useful for agents to decide which sections to load.
 */
export function listSkillSections(
  parsed: ParsedSkillContent
): Array<{ path: string; level: number; tokenEstimate: number }> {
  const result: Array<{ path: string; level: number; tokenEstimate: number }> = [];

  function walk(section: SkillSection): void {
    result.push({
      path: section.path,
      level: section.level,
      tokenEstimate: section.tokenEstimate,
    });
    for (const child of section.children) {
      walk(child);
    }
  }

  for (const section of parsed.sections) {
    walk(section);
  }

  return result;
}
